src/semantic/mod.rs

```diff
@@ -150,10 +150,18 @@ impl SemanticChecker {
         }
     }
 
-    fn check_block(&self, type_env: &mut TypeEnv, b: &Block, return_type: &Type) -> Result<()> {
+    fn check_block(&self, type_env: &TypeEnv, b: &Block, return_type: &Type) -> Result<()> {
```

Comment:
`type_env` became `&` not `&mut` since always create a new one for each block. Upper env only needs for lookup, don't need mutable.